Add docfx documentation

nitanmarcel opened the pull request:
An year-ish ago I've made a small issue speaking about a documentation website for the library generated using Doxygen https://github.com/crosire/scripthookvdotnet/issues/918 and since then I forgot about it, oops.

Anyway, using Doxygen for the documentation wasn't such a good idea since the generated HTML files needed some changes to display as they should be. 

A couple of days ago I found DocFx and I remembered about the documentation I've made for the library before.

The documentation is generated by DocFx each time a new tag is released which means it won't be available after merging it unless there's a new tag after the merge (the tag needs to start with v3 and also the gh-pages branch needs to be set as the main branch for GitHub pages) . But you can still test the theme here: https://nitanmarcel.github.io/scripthookvdotnet/index.html until it will be on this repo.


I'm using GitHub Actions to set up the documentation and it works in the following way:

If a new tag who's string matches "v3*" (eg. v3.1.1, v3.1.2, v3.x.x) will be release the workflow will start generating the documentation.

I tried to auto-generate everything that I could auto-generate starting from the index page which uses the README file, up to the changelog page which uses https://github.com/heinrichreimer/action-github-changelog-generator which generates changelog data between each available tag in a nicely printed way: eg. https://github.com/skywinder/ActionSheetPicker-3.0/blob/develop/CHANGELOG.md

The workflow will take care of everything except for deploying the GitHub page for the first time. It still requires a one-time activation from GitHub's repository settings. But after the workflow finished pushing the documentation to the gh-pages branch.

![image](https://user-images.githubusercontent.com/41646249/109579360-6be0a400-7b01-11eb-8502-5d6c2dbab430.png)

![image](https://user-images.githubusercontent.com/41646249/109579445-8fa3ea00-7b01-11eb-9929-50cf74f71e41.png)

![image](https://user-images.githubusercontent.com/41646249/109579583-d691df80-7b01-11eb-82b8-c74ddd5f1857.png)

nitanmarcel committed:
Add docfx documentation

crosire commented on doc/docfx.json:
```diff
@@ -0,0 +1,77 @@
+        "fileMetadataFiles": [],
+        "globalMetadata": {
+            "_appTitle": "ScriptHookVDotNet",
+            "__appFooter": "Copyright (C) 2015 crosire",
```
`__appFooter`, `_gitContribute` and `_appLogoPath` appear to be unused?

nitanmarcel replied:
Should be used, well if I look now the footer isn't used. Also I put a line on the logo path as an workaround to remove the original logo since I haven't found a proper GTA V svg logo.

About the _gitContribuite I don't know yet how it really works so improve this doc link is broken. The view source for works just fine

nitanmarcel replied:
The footer had one `_` more and it didn't registered. `_gitContribute` was used, but since I've disabled contribuitors it won't be used anymore. Also if it happens to have a svg logo for the site, I'll add it ^^ Until then I'll use the `-` as a way to remove it

crosire commented on doc/docfx.json:
```diff
@@ -0,0 +1,77 @@
+        ],
+        "dest": "_site",
+        "globalMetadataFiles": [],
+        "fileMetadataFiles": [],
```
Are these two entries necessary, consider that they are empty?

nitanmarcel replied:
I've took them from the documentation example and they were there. I'll see what happens if I remove them

crosire commented on doc/template/darkfx/partials/affix.tmpl.partial:
```diff
@@ -0,0 +1,39 @@
+      <ul class="nav">
+        {{#docurl}}
+        <li>
+          <a href="{{docurl}}" class="contribution-link">{{__global.improveThisDoc}}</a>
```
It looks like this is broken (at least on the demo in your repository). The link is displayed, but doesn't point anywhere, so may want to disable it by setting `_disableContribution`?

nitanmarcel replied:
That will break the view source code but thanks for highlight where the improve this doc part is ^^ I'll remove it from there

crosire reviewed (commented):
This is really cool!
Do you still have links to where the template used came from? And would it maybe be possible to embed that somehow without adding all the files to this repository (e.g. via a submodule or just fetching it as part of the actions script)?

nitanmarcel commented on doc/scripting_v3/index.md:
```diff
@@ -0,0 +1,4 @@
+ScriptHookVDotNet v2
```
Oh, my bad. I noticed this I had like over 30 commits and I kept reverting them when things went wrong and I might delete the one that modified this

nitanmarcel commented on doc/scripting_v2/index.md:
```diff
@@ -0,0 +1,5 @@
+ScriptHookVDotNet v2
+==========================
+
+
```
Oh, I'll fix it soon

nitanmarcel reviewed (commented):
> This is really cool!
> Do you still have links to where the template used came from? And would it maybe be possible to embed that somehow without adding all the files to this repository (e.g. via a submodule or just fetching it as part of the actions script)?

https://github.com/steffen-wilke/darkfx

In the second part you mean without adding the doc directory here? I could simply make my own repo with the doc folder and clone it using GitHub actions

nitanmarcel committed:
Fix index.md in doc/scripting_v2 and doc/scripting_v2

nitanmarcel committed:
Remove unused docfx.json keys

nitanmarcel committed:
Fix footer

nitanmarcel committed:
Disable Contribution links

nitanmarcel committed:
Change app's title

nitanmarcel commented:
@crosire I'm preparing a separate branch for the documentation so you can add it as a submodule if you want or I'll clone it from inside github actions. I'll wait for an answer ^^

nitanmarcel committed:
Use the documentation as a submodule

nitanmarcel committed:
Also checkout submodules

nitanmarcel commented:
I've added the documentation as a submodule: https://github.com/nitanmarcel/scripthookvdotnet-doc which is a fork of the template used on the documentation.

Also I did some changes. The main page uses the __Getting Started__ wiki page and the footer contains a link to the license which now will be generated as part of the website

nitanmarcel committed:
Update documentation submodule

Added some more generated filesto .gitignore in case the documentation is generated locally

crosire commented on doc/docfx.json:
```diff
@@ -0,0 +1,77 @@
+                    "scripting_v2/**.yml",
+                    "scripting_v3/**.yml",
+                    "scripting_v2/api/**.yml",
+                    "scripting_v3/api**.yml"
```
I mean you have one line
`"scripting_v2/api/**.yml"`
and then
`"scripting_v3/api**.yml"`
which is inconsistent?

crosire commented:
Ah, sorry :upside_down_face:, I didn't mean it that drastic, was thinking that the template could have been a submodule to avoid adding all these files that have no relation to SHVDN. So a directory structure like this:
```
- docs
   - docfx.json
   - index.md
   - toc.yml
   - scripting_v2/
      - index.md
   - scripting_v3/
      - index.md
   - templates/
      - darkfx/ submodule pointing towards https://github.com/steffen-wilke/darkfx (since you can have Git submodules in sub directories, see "-path" argument to git submodule add)
      - memberpage/ submodule pointing towards where that came from
```
While adding the entire docs as a submodule may be advantageous too, the problem I'm seeing is that it breaks the link between the GitHub Actions script and the docs content. So would be better to keep those close together I think.

nitanmarcel commented:
> Ah, sorry 🙃, I didn't mean it that drastic, was thinking that the template could have been a submodule to avoid adding all these files that have no relation to SHVDN. So a directory structure like this:
> 
> ```
> - docs
>    - docfx.json
>    - index.md
>    - toc.yml
>    - scripting_v2/
>       - index.md
>    - scripting_v3/
>       - index.md
>    - templates/
>       - darkfx/ submodule pointing towards https://github.com/steffen-wilke/darkfx (since you can have Git submodules in sub directories, see "-path" argument to git submodule add)
>       - memberpage/ submodule pointing towards where that came from
> ```
> 
> While adding the entire docs as a submodule may be advantageous too, the problem I'm seeing is that it breaks the link between the GitHub Actions script and the docs content. So would be better to keep those close together I think.

Ok, I'll see what I can do about it. Also the memberspage is a nuget package and I don't think it's built in in docfx like the default template but I will check :)

ChemsoDev commented:
Awesome work!!!

nitanmarcel commented:
> Awesome work!!!

Thanks. Anyway this will be pushed forward because I'm waiting for docFX v3 to come stable

kagikn commented:
I didn't care about web documentation too much until I received full access to do what I want without external approvals, but now it's good time for me to make a proper web documentation since we'll officially deprecate the v2 SDK/API in v3.7.0 (3.5 year passed since the first v3 release though). I'll make another PR when I create the official web docs.

I heard Lemon got annoyed for not a few people asking for how to use LemonUI in scripts built against v2 API (not v3), I'm disappointed the fact a lot of people (maybe more than the half of script devs on 5mods?) use v2 SDK/API for new scripts even they don't reference to other scripts built against v2 API. I'll be willing to create the official web documentation for API(s) if more people will have their scripts migrate to v3 SDK/API.

nitanmarcel commented:
I totally forgot this existed. Oups xD
